### PR TITLE
controku: init at 1.1.0

### PR DESCRIPTION
**Pull request by mjm**
## Description of changes

Add package for [controku](https://github.com/benthetechguy/), a Python library and GTK3 app for controlling Roku TVs.

Also add `ssdpy`, a dependency library used to discover TVs on the network.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by AndersonTorres** [comment]
Restart.

**Comment by mjm** [comment]
What does "restart" mean? This is the first new package I've added, did I do something wrong?

**Comment by AndersonTorres** [comment]
Originally you marked "fits CONTRIBUTING", while it obviously does not fit CONTRIBUTING in any way.

**Comment by mjm** [comment]
I see, I had read through CONTRIBUTING but had missed that specific information. I see it there now, and will fix it before marking it ready again.

You say it doesn't fit CONTRIBUTING in any way: is there more that I've missed? I guess maybe there's formatting changes needed, I'll look at that too. Like I said, this is my first time adding a new package: what's obvious to you is probably not obvious to me.

**Comment by mjm** [comment]
I think this is ready? Please let me know if I missed something.

File: pkgs/development/python-modules/ssdpy/default.nix [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  };
+
+  nativeBuildInputs = [ setuptools ];
+
+  meta = with lib; {
+    changelog = "https://github.com/MoshiBin/ssdpy/releases/tag/${version}";
+    description = "A lightweight, compatible SSDP library for Python.";
+    homepage = "https://github.com/MoshiBin/ssdpy";
+    license = licenses.mit;
+    maintainers = [ ];
```
Comment (AndersonTorres) [inline pkgs/development/python-modules/ssdpy/default.nix:24]: You don't want to maintain this?

Comment (mjm) [reply pkgs/development/python-modules/ssdpy/default.nix:24]: I will maintain it if that's what's required, but I don't really know anything about the library. I just needed it as a dependency of controku that wasn't already packaged.

File: pkgs/development/python-modules/ssdpy/default.nix [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  src = fetchPypi {
+    inherit pname version;
+    hash = "sha256-SXHGoBD3fxR+zb7JWT8tkYfD+2Nliw9ewI9Me+I4dCU=";
+  };
+
+  nativeBuildInputs = [ setuptools ];
+
+  meta = with lib; {
+    changelog = "https://github.com/MoshiBin/ssdpy/releases/tag/${version}";
+    description = "A lightweight, compatible SSDP library for Python.";
```
Comment (AndersonTorres) [inline pkgs/development/python-modules/ssdpy/default.nix:21]:
```suggestion
    description = "A lightweight, compatible SSDP library for Python";
```

File: pkgs/development/python-modules/ssdpy/default.nix (hunk 1 above)
Comment (AndersonTorres) [reply pkgs/development/python-modules/ssdpy/default.nix:24]: Usually there is no need to be an expert to maintain a package. 
Further, if this lib is a dependency of your main target (controku), it comes to reason you are interested in keep it up-to-date with that target.

File: pkgs/by-name/co/controku/package.nix [hunk 3]
```diff
@@ -0,0 +1,11 @@
+{ python3Packages
+, wrapGAppsHook
+, gobject-introspection
+, gtk3
+,
```
Comment (AndersonTorres) [inline pkgs/by-name/co/controku/package.nix:5]:
```suggestion
```

File: pkgs/development/python-modules/controku/default.nix [hunk 4]
```diff
@@ -0,0 +1,32 @@
+{ lib
+, python3Packages
+, fetchFromGitHub
+, setuptools
+, requests
+, ssdpy
+,
```
Comment (AndersonTorres) [inline pkgs/development/python-modules/controku/default.nix:7]:
```suggestion
```

File: pkgs/by-name/co/controku/package.nix [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+}:
```
Comment (AndersonTorres) [inline pkgs/by-name/co/controku/package.nix:6]:
```suggestion
}:

```

File: pkgs/development/python-modules/controku/default.nix [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+}:
```
Comment (AndersonTorres) [inline pkgs/development/python-modules/controku/default.nix:8]:
```suggestion
}:

```

File: pkgs/development/python-modules/ssdpy/default.nix (hunk 1 above)
Comment (mjm) [reply pkgs/development/python-modules/ssdpy/default.nix:24]: Fair point, I'll add myself to it then.

**Comment by AndersonTorres** [comment]
Now it looks OK. However I have few expertise on Python packaging.

If no one else complain about this in three days, ping me and I will merge.

File: pkgs/by-name/co/controku/package.nix [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+, wrapGAppsHook
+, gobject-introspection
+, gtk3
+}:
+
+with python3Packages;
+toPythonApplication (controku.overrideAttrs (old: {
+  nativeBuildInputs = old.nativeBuildInputs ++ [ wrapGAppsHook gobject-introspection ];
+  propagatedBuildInputs = old.propagatedBuildInputs ++ [ gtk3 appdirs pygobject3 ];
+}))
```
Comment (AndersonTorres) [inline pkgs/by-name/co/controku/package.nix:11]:
```suggestion
{ python3Packages
}:

python3Packages.toPythonApplication (controku.override { buildApplication = true; })
```

File: pkgs/development/python-modules/controku/default.nix [hunk 8]
```diff
@@ -0,0 +1,32 @@
+{ lib
+, python3Packages
+, fetchFromGitHub
+, setuptools
+, requests
+, ssdpy
+}:
```
Comment (AndersonTorres) [inline pkgs/development/python-modules/controku/default.nix:7]:
```suggestion
, appdirs 
, pygobject3
, gobject-introspection
, gtk3 
, wrapGAppsHook 
, buildApplication ? false
}:
```

File: pkgs/development/python-modules/controku/default.nix [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+  format = "pyproject";
+
+  src = fetchFromGitHub {
+    owner = "benthetechguy";
+    repo = "controku";
+    rev = version;
+    hash = "sha256-sye2GtL3a77pygllZc6ylaIP7faPb+NFbyKKyqJzIXw=";
+  };
+
+  nativeBuildInputs = [ setuptools ];
```
Comment (AndersonTorres) [inline pkgs/development/python-modules/controku/default.nix:21]:
```suggestion
  nativeBuildInputs = [ 
    setuptools 
  ] ++ lib.optionals buildApplication [
    gobject-introspection 
    wrapGAppsHook 
  ];
```

File: pkgs/development/python-modules/controku/default.nix [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+
+  propagatedBuildInputs = [ requests ssdpy ];
```
Comment (AndersonTorres) [inline pkgs/development/python-modules/controku/default.nix:23]:
```suggestion
  propagatedBuildInputs = [ 
    requests 
    ssdpy 
  ] ++ lib.optionals buildApplication [
    gtk3 
    appdirs 
    pygobject3 
  ];
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]
I believe we can do this cleaner and better.
Let's move all the building logic to the main file, whereas the app will be just a thin wrapper.

(For the sake of test, do not force-push this yet.)

**Review by AndersonTorres — APPROVED** [review]

**Comment by AndersonTorres** [comment]
I have no hardware to test this, but the GUI worked nice.

**Comment by mjm** [comment]
I tested it with my Roku TV, including with these latest changes. :+1:

File: pkgs/development/python-modules/ssdpy/default.nix [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+  pname = "ssdpy";
+  version = "0.4.1";
+  format = "pyproject";
+
+  src = fetchPypi {
+    inherit pname version;
+    hash = "sha256-SXHGoBD3fxR+zb7JWT8tkYfD+2Nliw9ewI9Me+I4dCU=";
+  };
+
+  nativeBuildInputs = [ setuptools ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/ssdpy/default.nix:17]: This is missing a pythonImportsCheck

File: pkgs/development/python-modules/controku/default.nix [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+
+  propagatedBuildInputs = [
+    requests
+    ssdpy
+  ] ++ lib.optionals buildApplication [
+    gtk3
+    appdirs
+    pygobject3
+  ];
+
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/controku/default.nix:42]: This is missing a pythonImportsCheck

Comment (mjm) [reply pkgs/development/python-modules/controku/default.nix:42]: Added!

File: pkgs/development/python-modules/ssdpy/default.nix (hunk 11 above)
Comment (mjm) [reply pkgs/development/python-modules/ssdpy/default.nix:21]: Added the imports check!

To be able to run the tests, I had to change to fetching the sources from GitHub, as the sources from PyPI don't include the tests.

I also had to disable several tests that try to do networking. Let me know if there's a way to be able to run these that I'm not aware of.

Comment (AndersonTorres) [reply pkgs/development/python-modules/ssdpy/default.nix:21]: Disabling tests that require networking access is the general procedure around Nixpkgs.

File: pkgs/development/python-modules/ssdpy/default.nix [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+
+  nativeCheckInputs = [
+    pytestCheckHook
+    pytest-mock
+  ];
+
+  pythonImportsCheck = [ "ssdpy" ];
+
+  disabledTests = [
```
Comment (AndersonTorres) [inline pkgs/development/python-modules/ssdpy/default.nix:30]:
```suggestion
  disabledTests = [
    # They all require network access
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]
Please rebase the commits to follow the guidelines (basically squash all the fixups as they never were done to begin with...)

**Review by AndersonTorres — APPROVED** [review]

**Pull request merged**